numpy/random/mtrand/mtrand.pyx

```diff
@@ -5064,7 +5064,8 @@ cdef class RandomState:
             x_ptr = <char*><size_t>x.ctypes.data
             stride = x.strides[0]
             itemsize = x.dtype.itemsize
-            buf = np.empty_like(x[0])  # GC'd at function exit
+            # Ensure that buf has the same dtype as x
+            buf = np.empty_like(x[:1])  # GC'd at function exit
```

Comment:
I agree with both of you with the comment issue. I'll update it.

However @seberg you are right to be worried about the object case, looks like it was working before (in 1.11) but this fix breaks it for a reason that is not entirely obvious to me just yet.